File: zoomx/src/main/java/com/zoomx/zoomx/ui/requestlist/RequestAdapter.java
package com.zoomx.zoomx.ui.requestlist;

import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;

import com.zoomx.zoomx.R;
import com.zoomx.zoomx.model.RequestEntity;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Ibrahim AbdelGawad on 12/3/2017.
 */

public class RequestAdapter extends RecyclerView.Adapter<RequestViewHolder> implements Filterable {

    private List<RequestEntity> requests = new ArrayList<>();
    private OnRequestItemClickListener onRequestItemClickListener;
    private List<RequestEntity> mFilteredList;

    public RequestAdapter() {
    }

    public void setRequestEntityList(List<RequestEntity> requestEntityList, OnRequestItemClickListener
            onRequestItemClickListener) {
        this.requests = requestEntityList;
        mFilteredList = requestEntityList;
        this.onRequestItemClickListener = onRequestItemClickListener;
        notifyDataSetChanged();
    }

    @Override
    public RequestViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.request_item_row, parent, false);

        return new RequestViewHolder(view);
    }

    @Override
    public void onBindViewHolder(RequestViewHolder holder, int position) {
        holder.bind(this.mFilteredList.get(position), onRequestItemClickListener);
    }

    @Override
    public int getItemCount() {
        if (mFilteredList == null)
            return 0;
        return mFilteredList.size();
    }

    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence constraint) {
                List<RequestEntity> filteredList = new ArrayList<>();

                if (TextUtils.isEmpty(constraint)) {
                    filteredList = requests;
                } else {
                    for (RequestEntity request : requests) {
                        if ((String.valueOf(request.getCode())).equals(constraint)) {

                            filteredList.add(request);
                        }
                    }
                }

                FilterResults filterResults = new FilterResults();
                filterResults.values = filteredList;
                return filterResults;
            }

            @Override
            protected void publishResults(CharSequence constraint, FilterResults results) {
                mFilteredList = (List<RequestEntity>) results.values;
                notifyDataSetChanged();
            }
        };
    }

    public interface OnRequestItemClickListener {
        void onItemClick(RequestEntity requestEntity);
    }
}

File: zoomx/src/main/java/com/zoomx/zoomx/ui/requestlist/RequestAdapter.kt
package com.zoomx.zoomx.ui.requestlist

import android.support.v7.widget.RecyclerView
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import com.zoomx.zoomx.R
import com.zoomx.zoomx.model.RequestEntity
import java.util.*

/**
 * Created by Ibrahim AbdelGawad on 12/3/2017.
 */

class RequestAdapter : RecyclerView.Adapter<RequestViewHolder>(), Filterable {

    private var requests: MutableList<RequestEntity> = ArrayList()
    private var onRequestItemClickListener: OnRequestItemClickListener? = null
    private var mFilteredList: List<RequestEntity>? = null

    fun setRequestEntityList(requestEntityList: MutableList<RequestEntity>, onRequestItemClickListener: OnRequestItemClickListener) {
        this.requests = requestEntityList
        mFilteredList = requestEntityList
        this.onRequestItemClickListener = onRequestItemClickListener
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RequestViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.request_item_row, parent, false)

        return RequestViewHolder(view)
    }

    override fun onBindViewHolder(holder: RequestViewHolder, position: Int) {
        holder.bind(this.mFilteredList!![position], onRequestItemClickListener)
    }

    override fun getItemCount(): Int {
        return if (mFilteredList == null) 0 else mFilteredList!!.size
    }

    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(constraint: CharSequence): Filter.FilterResults {
                var filteredList: MutableList<RequestEntity> = ArrayList()

                if (TextUtils.isEmpty(constraint)) {
                    filteredList = requests
                } else {
                    for (request in requests) {
                        if (request.code.toString() == constraint) {

                            filteredList.add(request)
                        }
                    }
                }

                val filterResults = Filter.FilterResults()
                filterResults.values = filteredList
                return filterResults
            }

            override fun publishResults(constraint: CharSequence, results: Filter.FilterResults) {
                mFilteredList = results.values as List<RequestEntity>
                notifyDataSetChanged()
            }
        }
    }

    interface OnRequestItemClickListener {
        fun onItemClick(requestEntity: RequestEntity)
    }
}
